app/inpututils.cpp

```diff
@@ -1842,57 +1842,84 @@ QgsGeometry InputUtils::createGeometryForLayer( QgsVectorLayer *layer )
     return geometry;
   }
 
-  switch ( layer->wkbType() )
+  bool isMulti = QgsWkbTypes::isMultiType( layer->wkbType() );
+
+  switch ( layer->geometryType() )
   {
-    case QgsWkbTypes::Point:
+    case QgsWkbTypes::PointGeometry:
     {
-      QgsPoint *point = new QgsPoint();
-      geometry.set( point );
+      if ( isMulti )
+      {
+        QgsMultiPoint *multiPoint = new QgsMultiPoint();
+        geometry.set( multiPoint );
+      }
+      else
+      {
+        QgsPoint *point = new QgsPoint();
+        geometry.set( point );
+      }
       break;
     }
 
-    case QgsWkbTypes::LineString:
+    case QgsWkbTypes::LineGeometry:
     {
-      QgsLineString *line = new QgsLineString();
-      geometry.set( line );
+      if ( isMulti )
+      {
+        QgsMultiLineString *multiLine = new QgsMultiLineString();
+        geometry.set( multiLine );
+      }
+      else
+      {
+        QgsLineString *line = new QgsLineString();
+        geometry.set( line );
+      }
       break;
     }
 
-    case QgsWkbTypes::Polygon:
+    case QgsWkbTypes::PolygonGeometry:
     {
-      QgsLineString *line = new QgsLineString();
-      QgsPolygon *polygon = new QgsPolygon( line );
-      geometry.set( polygon );
+      if ( isMulti )
+      {
+        QgsLineString *line = new QgsLineString();
+        QgsPolygon *polygon = new QgsPolygon( line );
+        QgsMultiPolygon *multiPolygon = new QgsMultiPolygon();
+        multiPolygon->addGeometry( polygon );
+        geometry.set( multiPolygon );
+      }
+      else
+      {
+        QgsLineString *line = new QgsLineString();
+        QgsPolygon *polygon = new QgsPolygon( line );
+        geometry.set( polygon );
+      }
       break;
     }
 
-    case QgsWkbTypes::MultiPoint:
-    {
-      QgsMultiPoint *multiPoint = new QgsMultiPoint();
-      geometry.set( multiPoint );
+    default:
       break;
-    }
+  }
 
-    case QgsWkbTypes::MultiLineString:
-    {
-      QgsMultiLineString *multiLine = new QgsMultiLineString();
-      geometry.set( multiLine );
-      break;
-    }
+  return geometry;
+}
 
-    case QgsWkbTypes::MultiPolygon:
-    {
-      QgsLineString *line = new QgsLineString();
-      QgsPolygon *polygon = new QgsPolygon( line );
-      QgsMultiPolygon *multiPolygon = new QgsMultiPolygon();
-      multiPolygon->addGeometry( polygon );
-      geometry.set( multiPolygon );
-      break;
-    }
 
-    default:
-      break;
+QString InputUtils::invalidGeometryWarning( QgsVectorLayer *layer )
+{
+  QString msg;
+  if ( !layer )
+  {
+    return msg;
   }
 
-  return geometry;
+  int nPoints = 1;
+  if ( layer->geometryType() == QgsWkbTypes::LineGeometry )
+  {
+    nPoints = 2;
+  }
+  else if ( layer->geometryType() == QgsWkbTypes::PolygonGeometry )
+  {
+    nPoints = 3;
+  }
+
+  return tr( "You need to add at least %1 point(s)." ).arg( nPoints );
```

Comment:
Can you please add a check there for multipart geometry and print `You need to add at least %1 point(s) to every part.` if it is multipart?